### ISSUE TITLE
amp-mustache: "failed to call initLogConstructor" error

### ISSUE THREAD

**Issue by BurakDev** [ampproject/amphtml#24390]
## What's the issue?

Our website randomly receive mixed version of AMP scripts and causing random errors on multiple pages/extensions.

Example 1 (working but with errors) :
![2019-09-06_13h51_35](https://user-images.githubusercontent.com/3963184/64426042-c8e6cf80-d0ad-11e9-98cd-5f34ccbcea62.png)

Example 1 (not working) :
![2019-09-06_13h51_00](https://user-images.githubusercontent.com/3963184/64426051-d56b2800-d0ad-11e9-9f4e-24ce17b0fe61.png)

Example 2 (working) :
![2019-09-06_13h52_16](https://user-images.githubusercontent.com/3963184/64426090-f5025080-d0ad-11e9-9cc7-ef607ed3c7ae.png)

Example 2 (not working) :
![2019-09-06_13h52_39](https://user-images.githubusercontent.com/3963184/64426113-03e90300-d0ae-11e9-8491-3e9dfbc16f37.png)

According to Chrome Network inspector, I receive both 1909051756330 and 1909051800160 revision of AMP scripts.
When I target single release (1909051756330 or 1909051800160), I don't get any error.

## How do we reproduce the issue?

Empty your cache and go on https://dictionary-beta.cambridge.org/plus/,
the wordlists list is randomly broken depending the versions you get from Google CDN.

amp-list randomly broken on our custom autocomplete implementation, at https://dictionary-beta.cambridge.org/, still depending of the versions you get from Google CDN.

To reproduce, you just need to clear your browser cache until you get the errors.

## What browsers are affected?

All browsers

## Which AMP version is affected?

1909051800160 and 1909051756330

**Comment by jridgewell** [comment]
The error itself seems to be caused by the rename of global variables from https://github.com/ampproject/amphtml/issues/24051. We're getting two different AMP versions loaded (`1909051800160` in amp-mustache.js and `1909051756330` in v0.js), where amp-mustache is referencing the old `self.logs` property and v0.js is referencing the new `self.__AMP_LOG` property. I still don't know how we loaded two different versions at the same time.

**Comment by choumx** [comment]
I think the first error is actually from `custom-elements-v1` experiment. Toggling the experiment off fixes it.

Check out the bottom few stacks:
```
v0.js:formatted:2449 Uncaught (in promise) TypeError: Cannot read property '__AMP_TOP' of null
    at de (v0.js:formatted:2449)
    at P (v0.js:formatted:2419)
    at Df (v0.js:formatted:3013)
    at nb.Ff (v0.js:formatted:3034)
    at new nb (amp-bind-0.1.js:28)
    at HTMLElement.Tj.b.createdCallback (v0.js:formatted:7081)
    at c.b [as constructor] (v0.js:formatted:7057)
    at new c (v0.js:formatted:7044)
    at Fa (v0.js:formatted:283)
    at Aa.f.upgradeSelf (v0.js:formatted:272)
```

**Comment by jridgewell** [comment]
It depends on if v0 loads canary or rc, but there are errors either way.

**Comment by jridgewell** [comment]
amp-mustache is able to load a bad version because of a flawed check in https://github.com/ampproject/amphtml/blob/ec0bc776c17c42eabbe50ebefb840c5f7d2e42db/src/runtime.js#L883-L885. It's looking for a script with `[custom-element="amp-mustache"]`, but amp-mustache is registered with `<script custom-template="amp-mustache">`

**Comment by choumx** [comment]
amp-mustache's wrong version appears to cause the second error:
```
v0.js:formatted:3698 Template failed:  [object HTMLTemplateElement] [object HTMLDivElement]: failed to call initLogConstructor
```

Separately, looks like this page is in the canary 1%. But amp-bind and amp-sidebar always have the wrong version (prod RTV). Why would only two extensions not be redirected to the canary RTV?

**Comment by choumx** [comment]
Extensions being diverted to different RTVs is currently expected behavior. May be fixed in the future.

**Comment by jridgewell** [comment]
Actually, something is injecting `https://cdn.ampproject.org/amp4ads-host-v0.js`, which is overriding `AMP_CONFIG`. @lannka

**Comment by jridgewell** [comment]
The only time I can see the first error is when `amp4ads-host-v0.js` is injected onto the page.

**Comment by BurakDev** [comment]
But why Google CDN is serving mixed AMP versions **ONLY** to visitors from our website ? (I checked with a script, this behavior happen only when there is our website http referer)

**Comment by jeffkaufman** [comment]
When I look at the source of https://dictionary-beta.cambridge.org/plus/ I see a strange mix of AMP and non-AMP.  For example, there's `<script async src="https://cdn.ampproject.org/v0.js"></script>` and other AMP scripts and elements, but there's also inline scripts and other things that are not valid AMP.

In this case it sounds like the problem is:

* One of your scripts is `gpt.js`, which is not expecting to run on a page with the AMP runtime
* `gpt.js` requests an ad, and gets an AMP ad in response
* When it writes the AMP ad to the page, the AMP ad is configured to use a different AMP runtime than the host page is using

Have you considered using AMP the standard way, with a page that passes AMP validation?

**Comment by jridgewell** [comment]
> But why Google CDN is serving mixed AMP versions ONLY to visitors from our website ?

You're being selected for 1% diversion based on the referring URL. The diversion selection then randomly chooses between multiple binaries (canary, release candidate, control). Since all the script requests are coming from the same referrer (your page), they all get selected for 1%. However, each request is separate, and each may choose a _different_ binary. We're working to make diversion selection consistent across requests, but don't have a timeline yet.

**Comment by BurakDev** [comment]
@jeffkaufman Yeah, but all of the bugs happen on AMP valid version too 😢 https://dictionary-beta.cambridge.org/plus/?forceAmp (which have a custom implementation of ads)
@jridgewell Yes I understand, that's why we are going to target single AMP release using https://cdn.ampproject.org/rtv/REVISION/ endpoint, we can't experiment untested revisions of AMP on our production platform

**Comment by jeffkaufman** [comment]
Thanks!  Visiting https://dictionary-beta.cambridge.org/plus/?forceAmp I no longer see see it loading gpt.js or amp4ads-host-v0.js so it does sound like the issue is related to a bad interaction with gpt.js after all, and my hypothesis was wrong.  Sorry!

**Comment by lannka** [comment]
I read through the issue, just to confirm that
1) `Cannot read property '__AMP_TOP' of null` is not the cause of the breakage. because the error occurs in your working example.
2) are you saying refreshing the page gives you different runtime version each time (either 1909051756330 or 1909051800160)? That will be really weird as we don't do 1% canary in canonical domain. Could you explain how to reproduce that? I tried cache clean + page refresh, but it did not reproduce that behavior. 

While debugging, one side thing I noticed, but likely is not relevant to the issue here, is that we have different cache-age for `v0.js` & `/v0/amp-*.js` which can cause `v0.js` to re-download the correct version of `/v0/amp-*.js`. That does seem like a version skew (different version of extension being loaded) but in fact by design and working as intended, should not cause the issue here.

**Comment by lannka** [comment]
hmm, @jeffkaufman your hypothesis might still be correct. it's possible that host.js is having a race condition with v0.js, and because AMP & inabox release is out sync, so mixed version is seen here.

**Comment by choumx** [comment]
The root bug is fixed in #24402.

**Comment by lannka** [comment]
Oh, cool. Shall we close this issue?

I think it worth a further investigation on @jeffkaufman hypothesis around host script. tracked separately: #24449
